std/experimental/checkedint.d

```diff
@@ -3063,3 +3105,90 @@ version(unittest) private struct CountOpBinary
     x = -1;
     assert(x > y);
 }
+
+// toHash
+@system unittest
+{
+    assert(checked(42).toHash() == checked(42).toHash());
+    assert(checked(12).toHash() != checked(19).toHash());
+
+    static struct Hook1
+    {
+        static size_t hookToHash(T)(T payload)
+        {
+            static if (size_t.sizeof == 4)
+            {
+                return typeid(payload).getHash(&payload) ^ 0xFFFF_FFFF;
+            }
+            else
+            {
+                return typeid(payload).getHash(&payload) ^ 0xFFFF_FFFF_FFFF_FFFF;
+            }
+
+        }
+    }
+
+    auto a = checked!Hook1(78);
+    auto b = checked!Hook1(78);
+    assert(a.toHash() == b.toHash());
+
+    assert(checked!Hook1(12).toHash() != checked!Hook1(13).toHash());
+
+    static struct Hook2
+    {
+        static if (size_t.sizeof == 4)
+        {
+            static size_t hashMask = 0xFFFF_0000;
+        }
+        else
+        {
+            static size_t hashMask = 0xFFFF_0000_FFFF_0000;
+        }
+
+        static size_t hookToHash(T)(T payload)
```

Comment:
Mark `hookToHash` as `@nogc nothrow @trusted` and then you can check that `toHash` correctly infers the attributes from his `Hook`. In order to check for the attributes add a `lambda` inside the assert, like so
`assert((() @nogc nothrow @safe => x.toHash() == x.toHash())());`

This defines a `lambda function` using the short syntax. You can read more about lambdas [here](https://dlang.org/spec/expression.html#lambdas)